File: src/main/java/edu/kit/iti/formal/automation/datatypes/promotion/IntegerPromotion.java
package edu.kit.iti.formal.automation.datatypes.promotion;

import edu.kit.iti.formal.automation.datatypes.Any;
import edu.kit.iti.formal.automation.datatypes.AnySignedInt;
import edu.kit.iti.formal.automation.datatypes.AnyUInt;

/**
 * Created by weigl on 24.11.16.
 */
public class IntegerPromotion implements TypePromotion {
    public static final IntegerPromotion INSTANCE = new IntegerPromotion();

    @Override
    public Any getPromotion(Any a, Any b) {
        try {
            return promote((AnySignedInt) a, (AnySignedInt) b);
        } catch (ClassCastException e) {
            try {
                return promote((AnyUInt) a, (AnyUInt) b);
            } catch (ClassCastException e1) {
                try {
                    return promote((AnySignedInt) a, (AnyUInt) b);
                } catch (ClassCastException e2) {
                    try{
                        return promote((AnySignedInt) b, (AnyUInt) a);
                    }catch (ClassCastException e3) {
                        return null;
                    }
                }
            }
        }
    }

    public Any promote(AnySignedInt a, AnySignedInt b) {
        if (a.getBitLength() >= b.getBitLength())
            return a;
        else
            return b;
    }


    public Any promote(AnyUInt a, AnyUInt b) {
        if (a.getBitLength() >= b.getBitLength())
            return a;
        else
            return b;
    }

    public Any promote(AnySignedInt a, AnyUInt b) {
        if (a.getBitLength() > b.getBitLength())
            return a;
        else
            return b.asSigned();
    }
}

File: iec61131lang/src/main/kotlin/edu/kit/iti/formal/automation/datatypes/promotion/IntegerPromotion.kt
package edu.kit.iti.formal.automation.datatypes.promotion

import edu.kit.iti.formal.automation.datatypes.AnyDt
import edu.kit.iti.formal.automation.datatypes.AnyInt


class IntegerPromotion : TypePromotion {
    override fun getPromotion(a: AnyDt, b: AnyDt) =
            if (a is AnyInt && b is AnyInt) {
                when {
                    (a.isSigned && b.isSigned) ->
                        if (a.bitLength >= b.bitLength) a else b
                    (!a.isSigned && !b.isSigned) ->
                        if (a.bitLength >= b.bitLength) a else b
                    else -> findNext(a, b)
                }
            } else null

    private fun findNext(a: AnyInt, b: AnyInt): AnyInt {
        var (signed, unsigned) = if (a.isSigned) Pair(a, b) else Pair(b, a)
        while (signed.upperBound < unsigned.upperBound) {
            if (signed.next().isPresent)
                signed = signed.next().get()
            else break
        }
        return signed
    }

    companion object {
        val INSTANCE = IntegerPromotion()
    }
}
